Use storage.v1 instead of v1beta1

jsafrane opened the pull request:
storage.v1beta1 was used to work around GKE which did not expose v1. Now that GKE is updated, we can switch everything to v1.

This is simple sed v1beta1 -> v1 + enabled a new test + changed preference of exposed interfaces in `storage/install/install.go`.

@msau42, PTAL and let me know when GKE is updated with storage v1 API and this PR can be actually merged.

@kubernetes/sig-storage-pr-reviews 

```release-note
NONE
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42477)
<!-- Reviewable:end -->

jsafrane commented:
BTW, does it require an exception?

msau42 commented:
This change LGTM.  I just made a GKE-side change that hopefully fixes the failing GKE tests (issue was not due to storageclass changes).  It'll probably take a few hours for the change to rollout to the test infra, but if you could rerun the GKE tests then, it should give us good confidence in this change.

msau42 commented:
@k8s-bot cvm gke e2e test this

msau42 commented:
@k8s-bot gci gke e2e test this

msau42 commented:
@k8s-bot gci gke e2e test this

msau42 commented:
@k8s-bot cvm gke e2e test this

jsafrane commented:
@msau42, I checked today with `KUBERNETES_PROVIDER=gke go run hack/e2e.go -v  --up` and GKE installed this way not expose storage.k8s.io/v1 API, thus tests in this PR fail.

```
$ kubectl version
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-02-15T06:34:56Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

msau42 commented:
Yes, because it's on version 1.5.3, while the default storage class is only allowed in 1.6+.  I'm pretty sure the Jenkins GKE tests exercise dynamic provisioning because earlier they were failing during the provisioning and now they are passing.  I'll try to find a dev cluster today so I can bring up 1.6 with your PR and run the storage tests.

msau42 commented on test/e2e/volume_provisioning.go:
```diff
@@ -103,48 +103,46 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 			class := newStorageClass("", "beta")
-			_, err := c.StorageV1beta1().StorageClasses().Create(class)
-			defer c.StorageV1beta1().StorageClasses().Delete(class.Name, nil)
+			_, err := c.StorageV1().StorageClasses().Create(class)
```
Should the beta tests still use beta APIs?

jsafrane replied:
Good idea, I added a new commit with that change, a bit more lines of code were necessary to use v1beta1 API.

msau42 commented:
I brought up a dev GKE cluster with your PR and ran all the dynamic provisioning tests + my new default sc ones (7 total) using the v1 API, and they all passed.

jsafrane commented:
@msau42, thanks for the tests. How are we going to proceed with this PR? I'd like to have it merged in 1.6, so the v1 API is actually used and tested there. AFAIK, GKE tests are disabled for now (or at least are non-blocking), so we can merge this PR, right?

jsafrane commented:
@k8s-bot unit test this
flake: #39471

@k8s-bot kops aws e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot cvm gce e2e test this

msau42 commented:
I agree we should try to get this into 1.6.  You're right that GKE tests are not queue-blocking anymore.  Regardless, I ran your PR on my dev GKE environment, so I am confident that they will not break this time :-)  And if they do, it'll be my job to fix them, not yours.

I pinged @ethernetdan and @idvoretskyi to see what they think.

kerneltime commented:
cc @divyenpatel there are changes related to vSphere tests, can you take a look.

divyenpatel commented:
We use e2e tests from master branch to validate 1.5 branch. so after this change, tests may fail to execute on new 1.5 based release?
cc: @BaluDontu

msau42 commented:
Yes, but how is that not a problem anytime you add a feature in a new release?

BaluDontu commented:
@msau42 : Its not just for the 1.5 releases but for all releases where v1beta1 is being used. For eg: 1.5 and 1.4 release branches. 

Since the storage class version will now change to v1 from v1beta1, the same tests which would now get executed on master branch would fail on 1.5, 1.4 release branch as v1 is not supported there? Is there any solution for this?

msau42 commented:
I feel like the best approach would be to run 1.5 version tests against 1.5, 1.4 against 1.4, etc.  This seems like a generic testing problem that is not specific to storage classes.  How is any new feature supposed to write tests for if we need to add versioning checks everywhere?

msau42 commented:
I spoke to @ethernetdan and he suggested that the PR be reduced to only just test changes.

@divyenpatel also explained the Vsphere testing environment.  They have volume tests checked into 1.6 but not previous branches, so they are running the 1.6 version of the tests against all releases.  I've asked them to backport their tests to each release instead.  For this PR, we can also remove the vsphere test changes for now until that happens, so the vsphere tests will still use the beta APIs.  Still, if the vsphere tests end up using some common wrapper/library that we changed, then the vsphere tests may still encounter issues.

erinboyd commented:
We should hold off on this until we have the ability to rev the tests as well to avoid having to revert this.

msau42 commented:
@erinboyd Do you see an issue with the main test framework, or just the vsphere ones?

erinboyd commented:
Discussed with with @msau42 and with vsphere approval and Jan's fix, it should be ready to go. @jeffvance @copejon @saad-ali

jsafrane commented:
Rebased, all e2e test changes were merged as #43296, so this PR is significantly smaller.

msau42 commented on pkg/apis/storage/install/install.go:
```diff
@@ -39,7 +39,7 @@ func Install(groupFactoryRegistry announced.APIGroupFactoryRegistry, registry *r
 		&announced.GroupMetaFactoryArgs{
 			GroupName: storage.GroupName,
 			// TODO:  change the order when GKE supports v1
```
This comment can be removed

jsafrane replied:
removed

msau42 commented on test/e2e/volume_provisioning.go:
```diff
@@ -107,7 +107,7 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
 
 	framework.KubeDescribe("DynamicProvisioner", func() {
 		It("should create and delete persistent volumes [Slow] [Volume]", func() {
-			framework.SkipUnlessProviderIs("openstack", "gce", "aws", "gke")
```
@divyenpatel is the vsphere CI ready, or should the vsphere changes be left out?

divyenpatel replied:
@msau42 Lets add vsphere to the supported provider list for this test.
For Internal vSphere CI, we can use local branch copy, until I cross port all our tests to respective active branches.

cc: @tusharnt, @BaluDontu

msau42 commented:
@k8s-bot gce etcd3 e2e test this

msau42 commented:
/lgtm

jsafrane commented:
@k8s-bot bazel test this

msau42 commented:
/lgtm

jsafrane commented:
@thockin, PTAL and approve

jsafrane commented:
rebased

msau42 commented:
/lgtm

jsafrane commented:
@thockin, please approve

thockin commented:
/approve

thockin commented:
@k8s-bot test this

msau42 commented:
W0424 13:55:30.870] # k8s.io/kubernetes/test/e2e/storage
W0424 13:55:30.871] test/e2e/storage/vsphere_volume_fstype.go:97: cannot use storageClassSpec (type *"k8s.io/kubernetes/pkg/apis/storage/v1".StorageClass) as type *"k8s.io/kubernetes/pkg/apis/storage/v1beta1".StorageClass in argument to client.StorageV1beta1().StorageClasses().Create
W0424 13:55:30.871] test/e2e/storage/vsphere_volume_fstype.go:101: cannot use storageclass (type *"k8s.io/kubernetes/pkg/apis/storage/v1beta1".StorageClass) as type *"k8s.io/kubernetes/pkg/apis/storage/v1".StorageClass in argument to getVSphereClaimSpecWithStorageClassAnnotation

jsafrane committed:
Use storage.v1 instead of v1beta1

storage.v1beta1 was used to work around GKE which does not expose v1. Now that
GKE is updated, we can switch everything to v1.

jsafrane committed:
Regenerate everything

jsafrane commented:
vsphere fixed

jsafrane commented:
@k8s-bot unit test this

jsafrane commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot non-cri e2e test this

jsafrane commented:
integration tests fixed (I hope)

msau42 commented:
/lgtm
